### PR TITLE
【hydra No.9】darcy2d

### PR DESCRIPTION
**Pull request by co63oc**
### PR types

Others 
### PR changes

Others 
### Describe

https://github.com/PaddlePaddle/PaddleScience/issues/571

9 darcy2d

### CODE REVIEW

**Comment by co63oc** [comment]
CI中按原参数测试会有错误
image

File: examples/darcy/conf/darcy2d.yaml [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+TRAIN:
+  epochs: 10000
+  iters_per_epoch: 1
+  lr_scheduler:
+    epochs: ${TRAIN.epochs}
+    iters_per_epoch: ${TRAIN.iters_per_epoch}
+    max_learning_rate: 1.0e-3
+    end_learning_rate: 1.0e-7
+  batch_size:
+    residual_validator: 8192
```
Comment (HydrogenSulfate) [inline examples/darcy/conf/darcy2d.yaml:49]: residual_validator移动到EVAL.batch_size

File: examples/darcy/conf/darcy2d.yaml [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    iters_per_epoch: ${TRAIN.iters_per_epoch}
+    output_dir: ./outputs_darcy2d_L-BFGS
+    learning_rate: 1.0
+    max_iter: 10
+    eval_freq: ${TRAIN.eval_freq}
+    eval_during_train: ${TRAIN.eval_during_train}
+
+# evaluation settings
+EVAL:
+  pretrained_model_path: null
```
Comment (HydrogenSulfate) [inline examples/darcy/conf/darcy2d.yaml:64]:
``` suggestion
eval_with_no_grad: false
```

File: examples/darcy/conf/darcy2d.yaml [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    epochs: ${TRAIN.epochs}
+    iters_per_epoch: ${TRAIN.iters_per_epoch}
+    max_learning_rate: 1.0e-3
+    end_learning_rate: 1.0e-7
+  batch_size:
+    residual_validator: 8192
+  eval_freq: 200
+  eval_during_train: true
+  pretrained_model_path: null
+  checkpoint_path: null
```
Comment (HydrogenSulfate) [inline examples/darcy/conf/darcy2d.yaml:53]: 移动至本级字段的最下方

File: examples/darcy/darcy2d.py [hunk 4]
```diff
@@ -205,18 +198,115 @@ def poisson_ref_compute_func(_in):
...
+    def poisson_ref_compute_func(_in):
+        return (
+            -8.0
+            * (np.pi**2)
+            * np.sin(2.0 * np.pi * _in["x"])
+            * np.cos(2.0 * np.pi * _in["y"])
+        )
+
+    # set validator
+    NPOINT_EVAL = cfg.NPOINT_PDE
```
Comment (HydrogenSulfate) [inline examples/darcy/darcy2d.py:227]: 这行可以删除

File: docs/zh/examples/darcy2d.md [hunk 5]
```diff
@@ -131,39 +143,39 @@ lambda _in: np.sin(2.0 * np.pi * _in["x"]) * np.cos(2.0 * np.pi * _in["y"])
...
+examples/darcy/darcy2d.py:78:82
 --8<--
 ```
 
 ### 3.5 超参数设定
 
 接下来我们需要指定训练轮数和学习率，此处我们按实验经验，使用一万轮训练轮数。
 
-``` py linenums="87"
+``` py linenums="41"
```
Comment (HydrogenSulfate) [inline docs/zh/examples/darcy2d.md:156]:
``` suggestion
``` yaml linenums="41"
```

**Review by HydrogenSulfate — CHANGES_REQUESTED** [review]

File: examples/darcy/darcy2d.py [hunk 6]
```diff
@@ -205,18 +198,115 @@ def poisson_ref_compute_func(_in):
...
+            "sampler": {"name": "BatchSampler"},
+        },
+        ppsci.loss.MSELoss("sum"),
+        evenly=True,
+        metric={"MSE": ppsci.metric.MSE()},
+        name="Residual",
+    )
+    validator = {residual_validator.name: residual_validator}
+
+    # set visualizer(optional)
```
Comment (HydrogenSulfate) [inline examples/darcy/darcy2d.py:245]:
``` suggestion
    # set visualizer
```

File: examples/darcy/darcy2d.py [hunk 7]
```diff
@@ -205,18 +198,115 @@ def poisson_ref_compute_func(_in):
...
+        output_dir=cfg.output_dir,
         equation=equation,
         geom=geom,
         validator=validator,
         visualizer=visualizer,
-        pretrained_model_path=f"{OUTPUT_DIR}/checkpoints/latest",
+        pretrained_model_path=cfg.EVAL.pretrained_model_path,
     )
     solver.eval()
     # visualize prediction for pretrained model(optional)
```
Comment (HydrogenSulfate) [inline examples/darcy/darcy2d.py:297]:
``` suggestion
    # visualize prediction
```

**Review by HydrogenSulfate — CHANGES_REQUESTED** [review]

File: docs/zh/examples/darcy2d.md [hunk 8]
```diff
@@ -185,19 +197,19 @@ examples/darcy/darcy2d.py:118:156
...
 
 #### 3.9.2 使用 L-BFGS 微调[可选]
 
 在使用 `Adam` 优化器训练完毕之后，我们可以将优化器更换成二阶优化器 `L-BFGS` 继续训练少量轮数（此处我们使用 `Adam` 优化轮数的 10% 即可），从而进一步提高模型精度。
 
-``` py linenums="181"
+``` py linenums="172"
 --8<--
-examples/darcy/darcy2d.py:181:206
+examples/darcy/darcy2d.py:172:192
```
Comment (HydrogenSulfate) [inline docs/zh/examples/darcy2d.md:212]: 172:198

File: docs/zh/examples/darcy2d.md [hunk 9]
```diff
@@ -131,39 +143,39 @@ lambda _in: np.sin(2.0 * np.pi * _in["x"]) * np.cos(2.0 * np.pi * _in["y"])
...
 
 接下来我们需要指定训练轮数和学习率，此处我们按实验经验，使用一万轮训练轮数。
 
-``` py linenums="87"
+``` yaml linenums="41"
 --8<--
-examples/darcy/darcy2d.py:87:88
+examples/darcy/conf/darcy2d.yaml:41:47
 --8<--
 ```
```
Comment (HydrogenSulfate) [inline docs/zh/examples/darcy2d.md:160]: 39:47

**Review by HydrogenSulfate — CHANGES_REQUESTED** [review]
辛苦修改之后顺便合入一下最新分支，里面解决了CI挂掉的问题

**Comment by co63oc** [comment]
已合并develop分支

**Comment by co63oc** [comment]
image
CI错误提示是tools目录
代码没有tools目录
image

**Review by HydrogenSulfate — APPROVED** [review]
LGTM

**Comment by pecanjk** [comment]
L-BFGS进行finetune时，loss nan，测过多次

图片

**Comment by HydrogenSulfate** [comment]
> L-BFGS进行finetune时，loss nan，测过多次
> 
> 图片

我这里运行没出现问题，ubuntu系统，paddlepaddle-develop + PaddleScience develop
image
image

**Comment by pecanjk** [comment]
我的是ubuntu 22.04 
NVIDIA-SMI 535.183.01             Driver Version: 535.183.01   CUDA Version: 12.2
paddlepaddle-gpu是cuda11.8版的

paddlepaddle-gpu         3.0.0.dev20250205
paddlesci                1.0.0.post607+g22231b5 /home/xxx/PaddleScience

我这Adam这一步最后的loss比你的高不少
图片

**Comment by HydrogenSulfate** [comment]
> 我的是ubuntu 22.04 NVIDIA-SMI 535.183.01 Driver Version: 535.183.01 CUDA Version: 12.2 paddlepaddle-gpu是cuda11.8版的
> 
> paddlepaddle-gpu 3.0.0.dev20250205 paddlesci 1.0.0.post607+g22231b5 /home/xxx/PaddleScience
> 
> 我这Adam这一步最后的loss比你的高不少 图片

可以尝试在AIStudio上新建一个项目运行试试看，可能是硬件问题？
